### PR TITLE
Sign the transaction in the InvokeContractVerify method

### PR DESCRIPTION
**Pull request by chenzhitong**
close https://github.com/neo-project/neo-modules/issues/423

### CODE REVIEW

**Review by shargon — COMMENTED** [review]
tested?

**Comment by superboyiii** [comment]
Wait for https://github.com/neo-project/neo/pull/2167

**Comment by superboyiii** [comment]
Let's move on.

File: src/RpcServer/RpcServer.Wallet.cs [hunk 1]
```diff
@@ -324,6 +325,56 @@ protected virtual JObject SendToAddress(JArray _params)
...
+            {
+                Signers = signers.GetSigners(),
+                Attributes = Array.Empty<TransactionAttribute>()
+            };
+            ContractParametersContext context = new ContractParametersContext(tx);
+            wallet.Sign(context);
+            tx.Witnesses = context.Completed ? context.GetWitnesses() : null;
+
+            using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Verification, tx, snapshot.Clone());
+            engine.LoadContract(contract, methodName, CallFlags.None, true, 0);
```
Comment (Tommo-L) [inline src/RpcServer/RpcServer.Wallet.cs:358]:
```suggestion
            engine.LoadContract(contract, methodName, CallFlags.None, true, args.length);
```

**Comment by Tommo-L** [comment]
Plz, wait https://github.com/neo-project/neo/pull/2186

**Comment by shargon** [comment]
Tested?

File: src/RpcServer/RpcServer.Wallet.cs [hunk 2]
```diff
@@ -324,6 +325,56 @@ protected virtual JObject SendToAddress(JArray _params)
...
+            {
+                Signers = signers.GetSigners(),
+                Attributes = Array.Empty<TransactionAttribute>()
+            };
+            ContractParametersContext context = new ContractParametersContext(tx);
+            wallet.Sign(context);
+            tx.Witnesses = context.Completed ? context.GetWitnesses() : null;
+
+            using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Verification, tx, snapshot.Clone());
+            engine.LoadContract(contract, methodName, CallFlags.None, true, args.length);
```
Comment (cloud8little) [inline src/RpcServer/RpcServer.Wallet.cs:358]:
```suggestion
            engine.LoadContract(contract, methodName, CallFlags.None, true, (ushort)args.Length);
```

File: src/RpcServer/RpcServer.Wallet.cs [hunk 3]
```diff
@@ -324,6 +325,56 @@ protected virtual JObject SendToAddress(JArray _params)
...
+            {
+                Signers = signers.GetSigners(),
+                Attributes = Array.Empty<TransactionAttribute>()
+            };
+            ContractParametersContext context = new ContractParametersContext(tx);
+            wallet.Sign(context);
+            tx.Witnesses = context.Completed ? context.GetWitnesses() : null;
+
+            using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Verification, tx, snapshot.Clone());
+            engine.LoadContract(contract, methodName, CallFlags.None, true, (ushort)args.Length);
```
Comment (cloud8little) [inline src/RpcServer/RpcServer.Wallet.cs:358]: thought there is no need for LoadContract here.

**Comment by cloud8little** [comment]
Test result: Fail   

- Invoke OracleContract with a random account, it should return false.

a2d16f6f64075a81fbd12bfd804298b

- Invoke a new deployed NEP17 contract with owner account, it should return true.
af7564cb2a902ff5e36d70b8cf55819

**Comment by cloud8little** [comment]
Conflict need to be resolved.

File: src/RpcServer/RpcServer.Wallet.cs (hunk 3 above)
Comment (shargon) [reply src/RpcServer/RpcServer.Wallet.cs]: Please check it again

File: src/RpcServer/RpcServer.Wallet.cs [hunk 4]
```diff
@@ -324,6 +325,54 @@ protected virtual JObject SendToAddress(JArray _params)
...
+            }
+            var methodName = "verify";
+
+            Transaction tx = signers == null ? null : new Transaction
+            {
+                Signers = signers.GetSigners(),
+                Attributes = Array.Empty<TransactionAttribute>()
+            };
+            ContractParametersContext context = new ContractParametersContext(tx);
+            wallet.Sign(context);
```
Comment (vncoelho) [inline src/RpcServer/RpcServer.Wallet.cs:354]: Is there a need to check if these `signers` are know or unknown as @cloud8little tested?

**Comment by cloud8little** [comment]
Test Passed: 

image

image

**Review by shargon — APPROVED** [review]

**Comment by Tommo-L** [comment]
@cloud8little Did you test it with more than one parameters? It seems it'll fail.

File: src/RpcServer/RpcServer.Wallet.cs [hunk 5]
```diff
@@ -324,6 +325,54 @@ protected virtual JObject SendToAddress(JArray _params)
...
+            tx.Witnesses = context.Completed ? context.GetWitnesses() : null;
+
+            using ApplicationEngine engine = ApplicationEngine.Create(TriggerType.Verification, tx, snapshot.Clone());
+            engine.LoadScript(new ScriptBuilder().EmitDynamicCall(scriptHash, methodName, true, args).ToArray(), (ushort)args.Length, 1);
```
Comment (superboyiii) [inline src/RpcServer/RpcServer.Wallet.cs:358]: Please delete `ReturnValue` in EmitDynamicCall
```suggestion
            engine.LoadScript(new ScriptBuilder().EmitDynamicCall(scriptHash, methodName, args).ToArray(), (ushort)args.Length, 1);
```

**Comment by superboyiii** [comment]
@Tommo-L I've tested this and seems multi params are OK.

Demo:
```
using Neo;
using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Services.Neo;
using System;
using System.ComponentModel;
using System.Numerics;

namespace Contract1
{
    [ManifestExtra("Author", "Neo")]
    [ManifestExtra("Email", "dev@neo.org")]
    [ManifestExtra("Description", "This is a contract example")]
    public class Contract1 : SmartContract
    {
        [DisplayName("Verify")]
        public static event Action<int, int, int> OnVerify;

        static readonly UInt160 Owner = "NUXLF99jdK1v5hsgjL7pwMr4Tm7RXrwK6W".ToScriptHash();

        private static bool IsOwner() => Runtime.CheckWitness(Owner);

        public static bool Verify(int a, int b)
        {
            int c = a + b;
            OnVerify(a, b, c);
            return IsOwner();
        }

        public static string test2()
        {
            return "test2";
        }

        public static void _deploy(bool update)
        {
            Storage.Put(Storage.CurrentContext, "this", "is");
        }

        public static void test3(int a)
        {
            int b = a + 2;
        }

        public static int testInt(int a)
        {
            return 1;
        }

        public static void testvoidinputstring(string a)
        {
            string b = a + "abc";
        }
    }
}
``` 

Request:
```
{
  "jsonrpc": "2.0",
  "method": "invokecontractverify",
  "params": [
    "0xb78526cff2721e611de552ebf4c500e30d966848",
    [
      {
        "type": "Integer",
        "value": "1"
      },
      {
        "type": "Integer",
        "value": "2"
      }
    ],
    [
      {
        "account": "NUXLF99jdK1v5hsgjL7pwMr4Tm7RXrwK6W",
        "scopes": "CalledByEntry"
      }
    ]
  ],
  "id": 3
}
```

Response:
```
{
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
        "script": "VwABQZv2Z84MBHRoaXMMAmlzU0HmPxiEQFYBDBRecSTBQp/CCMobYfKsNidhhQGa4GBADBRecSTBQp/CCMobYfKsNidhhQGa4EH4J+yMQFcBAAwFdGVzdDJwaEBXAQF4Ep5wQFcBARFwaEBXAQF4DANhYmOL2yhwQFcCAnh5nnB4eWhTE8AMBlZlcmlmeUGVAW9hNKBxaUA=",
        "state": "HALT",
        "gasconsumed": "0.0204747",
        "exception": null,
        "stack": [
            {
                "type": "Boolean",
                "value": true
            }
        ]
    }
}
```

**Review by superboyiii — CHANGES_REQUESTED** [review]
Please delete `ReturnValue` in EmitDynamicCall

File: src/RpcServer/RpcServer.Wallet.cs [hunk 6]
```diff
@@ -324,6 +325,54 @@ protected virtual JObject SendToAddress(JArray _params)
...
+
```
Comment (superboyiii) [inline src/RpcServer/RpcServer.Wallet.cs:359]:
```suggestion
            engine.LoadScript(new ScriptBuilder().EmitDynamicCall(scriptHash, methodName, args).ToArray(), (ushort)args.Length, 1);

```

**Review by superboyiii — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Pull request merged**